File: java/Heaps/KMostFrequentStringsMinHeap.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class KMostFrequentStringsMinHeap {
    public class Pair implements Comparable<Pair> {
        String str;
        int freq;
        public Pair(String str, int freq) {
            this.str = str;
            this.freq = freq;
        }
        // Since this is a min-heap comparator, we can use the same 
        // comparator as the one used in the max-heap, but reversing the 
        // inequality signs to invert the priority.
        @Override
        public int compareTo(Pair other) {
            if (this.freq == other.freq) {
                return other.str.compareTo(this.str);
            }
            return Integer.compare(this.freq, other.freq);
        }
    }

    public List<String> kMostFrequentStringsMinHeap(String[] strs, int k) {
        List<String> res = new ArrayList<>();
        Map<String, Integer> freqs = new HashMap<>();
        for (String str : strs) {
            freqs.put(str, freqs.getOrDefault(str, 0) + 1);
        }
        PriorityQueue<Pair> minHeap = new PriorityQueue<>();
        for (Map.Entry<String, Integer> entry : freqs.entrySet()) {
            minHeap.add(new Pair(entry.getKey(), entry.getValue()));
            // If heap size exceeds 'k', pop the lowest frequency string to 
            // ensure the heap only contains the 'k' most frequent words so 
            // far.
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        // Return the 'k' most frequent strings by popping the remaining 'k' 
        // strings from the heap. Since we're using a min-heap, we need to 
        // reverse the result after popping the elements to ensure the most 
        // frequent strings are listed first.
        while (!minHeap.isEmpty()) {
            res.add(minHeap.poll().str);
        }
        Collections.reverse(res);
        return res;
    }
}

File: kotlin/Heaps/KMostFrequentStringsMinHeap.kt
import java.util.PriorityQueue

data class Pair(val str: String, val freq: Int)

fun kMostFrequentStringsMinHeap(strs: List<String>, k: Int): List<String> {
    val freqs = strs.groupingBy { it }.eachCount()
    val minHeap =
            PriorityQueue<Pair> { a, b ->
                // Since this is a min-heap comparator, we can use the same
                // comparator as the one used in the max-heap, but reversing the
                // inequality signs to invert the priority.
                if (a.freq == b.freq) a.str.compareTo(b.str) else a.freq.compareTo(b.freq)
            }
    for ((str, freq) in freqs) {
        minHeap.add(Pair(str, freq))
        // If heap size exceeds 'k', pop the lowest frequency string to
        // ensure the heap only contains the 'k' most frequent words so
        // far.
        if (minHeap.size > k) {
            minHeap.poll()
        }
    }
    // Return the 'k' most frequent strings by popping the remaining 'k'
    // strings from the heap. Since we're using a min-heap, we need to
    // reverse the result after popping the elements to ensure the most
    // frequent strings are listed first.
    return List(k) { minHeap.poll().str }.reversed()
}
